curp/src/client/state.rs

```diff
@@ -0,0 +1,233 @@
+use std::{
+    cmp::Ordering,
+    collections::{hash_map::Entry, HashMap, HashSet},
+    sync::Arc,
+};
+
+use tokio::sync::RwLock;
+use tracing::debug;
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        self,
+        connect::{BypassedConnect, ConnectApi},
+        FetchClusterResponse, Protocol,
+    },
+};
+
+/// Reference to state
+pub(super) type StateRef = Arc<RwLock<State>>;
+
+/// Client state
+pub(super) struct State {
+    /// Leader id. At the beginning, we may not know who the leader is.
+    pub(super) leader: Option<ServerId>,
+    /// Local server id
+    pub(super) local_server: Option<ServerId>,
+    /// Term, initialize to 0, calibrated by the server.
+    pub(super) term: u64,
+    /// Cluster version, initialize to 0, calibrated by the server.
+    pub(super) cluster_version: u64,
+    /// Members' connect
+    pub(super) connects: HashMap<ServerId, Arc<dyn ConnectApi>>,
+}
+
+impl std::fmt::Debug for State {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("State")
+            .field("leader", &self.leader)
+            .field("local_server", &self.local_server)
+            .field("term", &self.term)
+            .field("cluster_version", &self.cluster_version)
+            .field("connects", &self.connects.keys())
+            .finish()
+    }
+}
+
+impl State {
+    /// For test
+    #[cfg(test)]
+    pub(super) fn new_ref(
+        connects: HashMap<ServerId, Arc<dyn ConnectApi>>,
+        local_server: Option<ServerId>,
+        leader: Option<ServerId>,
+        term: u64,
+        cluster_version: u64,
+    ) -> StateRef {
+        Arc::new(RwLock::new(Self {
+            leader,
+            local_server,
+            term,
+            cluster_version,
+            connects,
+        }))
+    }
+
+    /// Get the local connect
+    pub(super) fn local_connect(&self) -> Option<Arc<dyn ConnectApi>> {
+        let id = self.local_server?;
+        self.connects.get(&id).map(Arc::clone)
+    }
+
+    /// Update leader
+    pub(super) fn check_and_update_leader(
+        &mut self,
+        leader_id: Option<ServerId>,
+        term: u64,
+    ) -> bool {
+        match self.term.cmp(&term) {
+            Ordering::Less => {
+                // reset term only when the resp has leader id to prevent:
+                // If a server loses contact with its leader, it will update its term for election. Since other servers are all right, the election will not succeed.
+                // But if the client learns about the new term and updates its term to it, it will never get the true leader.
+                if let Some(new_leader_id) = leader_id {
+                    debug!("client term updates to {}", term);
+                    debug!("client leader id updates to {new_leader_id}");
```

Comment:
We can merge these two lines together, and use `info!` instead.

Reply:
done